File: norm_ls.lua
local json = require('json')
local norm_ls = {}

function norm_ls.format_error(error)
	return {
		range = {
			start = {
				line = error.line - 1,
				char = error.char - 1,
			},
			["end"] = {
			line = error.line,
			char = error.char,
			},
			severity = 1,
			code = error.code,
			codeDescription = error.message,
			source = "norm_ls",
			message = error.code,
		}
	}
end


--- capture the output of a command into a string
---@param cmd string command to be ran
---@param raw boolean wether to process output or not
---@return string s output of command
function os.capture(cmd, raw)
	local f = assert(io.popen(cmd, 'r'))
	local s = assert(f:read('*a'))
	f:close()
	if raw then return s end
	s = string.gsub(s, '^%s+', '')
	s = string.gsub(s, '%s+$' , '')
	s = string.gsub(s, '[\n\r]+', '')
	return s
end

function norm_ls.get_errors(file)
	local lines = os.capture("norminette " .. file , true):gmatch("Error: [^\n]+")
	local errors = {}

	for line in lines do
		table.insert(errors, {
			code = line:match("Error:%s*[%u_]+"):match("[%u_]+", 2),
			line = line:match("line:%s*%d+"):match("%d+"),
			char = line:match("col:%s*%d+"):match("%d+"),
			message = line:match("%):.*"):match("%u[%a%s]+")
		})
	end
	return errors
end

function norm_ls.get_diagnostics(file)
	local errors = norm_ls.get_errors(file)
	local diagnostics = {}

	for _,v in pairs(errors) do
		table.insert(diagnostics, norm_ls.format_error(v))
	end
	return diagnostics
end

function norm_ls.publish_diagnostics(diagnostics, file)
	local body = json.encode{
		jsonrpc = "2.0",
		method = "textDocument/publishDiagnostics",
		params = {
			uri = file,
			diagnostics = diagnostics
		}
	}
	io.write("Content-Length: " .. #body .. "\r\n\r\n" .. body)
	io.flush()
end

function norm_ls.publish_capabilities(id)
	local body = json.encode({
		jsonrpc = "2.0",
		id = id,
		result = {
			capabilities = {
				textDocumentSync = {
					{ openClose = true },
					{ change = 1 },
					{ save = { includeText = false } },
				},
				diagnosticProvider = {
					{ interFileDependencies = false },
					{ workspaceDiagnostics = false },
				}
			}
		}
	})
	io.write("Content-Length: " .. #body .. "\r\n\r\n" .. body)
	io.flush()
end

function norm_ls.read_input()
	local headers = {}
	while true do
		local line = io.read("*l")
		if not line then return nil end
		if line == "" then break end
		local name, value = line:match("^(.-):%s*(.+)$")
		if name then headers[name:lower()] = value else return nil end
	end

	local content_length = tonumber(headers["content-length"]) or 0
	if not content_length or content_length < 0 then
		return nil
	end

	local body = io.read(content_length)
	if not body or #body ~= content_length then return nil end

	return json.decode(body)
end

function norm_ls.handle_request(request)
	if request.method == "initialize" then
		norm_ls.publish_capabilities(request.id)
	elseif request.method == "textDocument/didOpen" or request.method == "textDocument/didChange" or request.method == "textDocument/didSave" then
		norm_ls.publish_diagnostics(norm_ls.get_diagnostics(request.file), request.file)
	end
end

while true do 
	local request = norm_ls.read_input()
	if request then
		norm_ls.handle_request(request)
	end
end

File: norm_ls.py
#!/usr/bin/env python3
import json
import logging
from subprocess import PIPE, Popen
from sys import stderr

from lsprotocol import types
from pygls import uris
from pygls.cli import start_server
from pygls.lsp.server import LanguageServer
from pygls.workspace import TextDocument


class Hint:
    def __init__(self, code: str, message: str, line: str, char: str) -> None:
        self.code = code
        self.message = message
        self.line = line
        self.char = char

    def __str__(self) -> str:
        return (
            f"[{self.line}:{self.char}]{self.code.capitalize()}"
            f": {self.message}"
        )

    def __repr__(self) -> str:
        return str(self)


def get_hints(file: str) -> list[Hint] | None:
    TIMEOUT: int = 5
    process = Popen(
        [
            "norminette",
            "-d",
            "-R",
            "CheckForbiddenSourceHeader",
            "-f",
            "json",
            file,
        ],
        stdout=PIPE,
    )
    try:
        (output, _) = process.communicate(timeout=TIMEOUT)
        _ = process.wait(timeout=TIMEOUT)
    except Exception as e:
        print(f"Failed to gather hints: {e}", file=stderr)
        process.kill()
    line = output.splitlines()[1]
    if not line:
        return None
    content = json.loads(line)["files"][0]
    status = content["status"]
    errors = content["errors"]
    hints: list[Hint] = []
    if status == "Error":
        for h in errors:
            hints.append(
                Hint(
                    h["name"],
                    h["text"],
                    h["highlights"][0]["lineno"],
                    h["highlights"][0]["column"],
                )
            )
    return hints


class PublishDiagnosticServer(LanguageServer):
    def __init__(self, *args, **kwargs) -> None:  # noqa: ANN002, ANN003
        super().__init__(*args, **kwargs)
        self.diagnostics = {}

    def parse(self, document: TextDocument) -> None:
        if not uris.to_fs_path(document.uri):
            return
        else:
            file = str(uris.to_fs_path(document.uri))
        hints = get_hints(file)
        diagnostics = []
        if hints:
            for hint in hints:
                diagnostics.append(
                    types.Diagnostic(
                        message=hint.message,
                        code=hint.code,
                        code_description=None,
                        severity=types.DiagnosticSeverity(value=1),
                        source="norm_ls",
                        range=types.Range(
                            start=types.Position(
                                line=int(hint.line) - 1,
                                character=int(hint.char) - 1,
                            ),
                            end=types.Position(
                                line=int(hint.line) - 1,
                                character=int(hint.char) - 1,
                            ),
                        ),
                    )
                )
        self.diagnostics[document.uri] = (document.version, diagnostics)


server = PublishDiagnosticServer(
    "norm_ls",
    "v0.1.3",
    text_document_sync_kind=types.TextDocumentSyncKind(1),
)


@server.feature(types.TEXT_DOCUMENT_DID_OPEN)
def did_open(
    ls: PublishDiagnosticServer, params: types.DidOpenTextDocumentParams
) -> None:
    doc = ls.workspace.get_text_document(params.text_document.uri)
    ls.parse(doc)

    for uri, (version, diagnostics) in ls.diagnostics.items():
        ls.text_document_publish_diagnostics(
            types.PublishDiagnosticsParams(
                uri=uri,
                version=version,
                diagnostics=diagnostics,
            )
        )


@server.feature(types.TEXT_DOCUMENT_DID_CHANGE)
def did_change(
    ls: PublishDiagnosticServer, params: types.DidChangeTextDocumentParams
) -> None:
    doc = ls.workspace.get_text_document(params.text_document.uri)
    ls.parse(doc)

    for uri, (version, diagnostics) in ls.diagnostics.items():
        ls.text_document_publish_diagnostics(
            types.PublishDiagnosticsParams(
                uri=uri,
                version=version,
                diagnostics=diagnostics,
            )
        )


@server.feature(types.TEXT_DOCUMENT_DID_SAVE)
def did_save(
    ls: PublishDiagnosticServer, params: types.DidSaveTextDocumentParams
) -> None:
    doc = ls.workspace.get_text_document(params.text_document.uri)
    ls.parse(doc)

    for uri, (version, diagnostics) in ls.diagnostics.items():
        ls.text_document_publish_diagnostics(
            types.PublishDiagnosticsParams(
                uri=uri,
                version=version,
                diagnostics=diagnostics,
            )
        )


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO, format="%(message)s")
    start_server(server)
